File: hub-workday-connector/src/test/java/com/vmware/ws1connectors/workday/web/CardsControllerTest.java
/*
 * Project Workday Connector
 * (c) 2019-2020 VMware, Inc. All rights reserved.
 * SPDX-License-Identifier: BSD-2-Clause
 */
package com.vmware.ws1connectors.workday.web;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;
import org.springframework.http.MediaType;

import java.io.IOException;
import java.util.stream.Collectors;

import static com.vmware.ws1connectors.workday.exceptions.ExceptionHandlers.INVALID_CONNECTOR_TOKEN;
import static com.vmware.ws1connectors.workday.utils.ApiUrlConstants.CARDS_REQUESTS_API;
import static com.vmware.ws1connectors.workday.utils.ApiUrlConstants.TIME_OFF_REQUEST_API_PATH;
import static net.javacrumbs.jsonunit.assertj.JsonAssertions.assertThatJson;
import static net.javacrumbs.jsonunit.core.Option.IGNORING_ARRAY_ORDER;
import static net.javacrumbs.jsonunit.core.Option.IGNORING_EXTRA_FIELDS;
import static org.hamcrest.Matchers.any;
import static org.springframework.http.HttpHeaders.ACCEPT;
import static org.springframework.http.HttpHeaders.AUTHORIZATION;
import static org.springframework.http.HttpMethod.GET;
import static org.springframework.http.HttpMethod.POST;
import static org.springframework.http.MediaType.APPLICATION_JSON;
import static org.springframework.test.web.client.match.MockRestRequestMatchers.header;
import static org.springframework.test.web.client.match.MockRestRequestMatchers.method;
import static org.springframework.test.web.client.match.MockRestRequestMatchers.requestTo;
import static org.springframework.test.web.client.response.MockRestResponseCreators.withSuccess;
import static org.springframework.test.web.client.response.MockRestResponseCreators.withUnauthorizedRequest;

public class CardsControllerTest extends ControllerTestUtils {
    private static final String UNAUTHORIZED_WORKDAY_AUTH_TOKEN = BEARER + "unauthorized-auth-token";
    private static final String NO_WORKDAY_AUTH_TOKEN = null;
    private static final String TIME_OFF_REQUEST_ID_1_API = TIME_OFF_REQUEST_API_PATH + "fc844b7a8f6f01580738a5ffd6115105";
    private static final String TIME_OFF_REQUEST_ID_2_API = TIME_OFF_REQUEST_API_PATH + "fc844b7a8f6f01223ce56358d5113e05";
    private static final String ERROR_JSON_PATH = "/error";
    private static final String BUSINESS_TITLE_CHANGE_URL = "/common/v1/businessTitleChanges/0db7bcdda1cd01fbdae2376a4844ab18";
    private static final String BUSINESS_PROCESS_URL = "/common/v1/businessProcesses/f387bb35571f017af0641c153a486314";
    private static final String BUSINESS_PROCESS_URL_1 = "/common/v1/businessProcesses/f387bb35571f013c819744b63a489314";
    private static final String BUSINESS_PROCESS_URL_WITH_STEP_TYPE_ACTION = "/common/v1/businessProcesses/4e7de5de24b7011d66c91c9a610cdb2d";
    private static final String BUSINESS_PROCESS_URL_WITH_STEP_TYPE_REVIEW_DOCUMENTS = "/common/v1/businessProcesses/318af1942cc001af54978e493e14fa0a";
    private static final String REQUEST_BODY_FILE_NAME = "get_cards_request_body.json";
    private static final String BUSINESS_PROCESS_URL_PRE_HIRE_1 = "/common/v1/businessProcesses/318af1942cc001af54978e493e14fa0a";
    private static final String BUSINESS_PROCESS_URL_PRE_HIRE_2 = "/common/v1/businessProcesses/318af1942cc001af76543e493e14fa0a";

    @Test public void testCardRequestsApiIsProtected() throws Exception {
        testProtectedResource(POST, CARDS_REQUESTS_API);
    }

    @Test public void testConnectorDiscoveryMetadata() throws IOException {
        testConnectorDiscovery();
    }

    @Test public void cardRequestFailsWhenUnauthorizedWorkdayAuthTokenIsProvided() throws IOException {
        mockBackend.expect(requestTo(any(String.class)))
            .andRespond(withUnauthorizedRequest());

        requestCards(UNAUTHORIZED_WORKDAY_AUTH_TOKEN, CARDS_REQUESTS_API, REQUEST_BODY_FILE_NAME, false)
            .expectStatus().isBadRequest()
            .expectBody().jsonPath(ERROR_JSON_PATH, INVALID_CONNECTOR_TOKEN);
    }

    @Test public void cardRequestFailsWhenWorkdayAuthTokenIsMissing() throws IOException {
        requestCards(NO_WORKDAY_AUTH_TOKEN, CARDS_REQUESTS_API, REQUEST_BODY_FILE_NAME, false)
            .expectStatus().isBadRequest();
    }

    @ParameterizedTest
    @ValueSource(strings = {"no_results.json", "inbox_task_without_status.json",
            "inbox_task_without_overall_process.json"})
    public void canGetEmptyCardsWhenNoTimeOffTasksFound(String inboxTaskFile) throws Exception {
        mockWorkdayApiResponse(getInboxTasksUri(), inboxTaskFile);

        requestCards(WORKDAY_AUTH_TOKEN, CARDS_REQUESTS_API, REQUEST_BODY_FILE_NAME, false)
            .expectStatus().isOk()
            .expectHeader().contentTypeCompatibleWith(APPLICATION_JSON)
            .expectBody().json(fromFile("cards/no_cards.json"));
    }

    @Test public void canGetCards() throws Exception {
        mockWorkdayApiResponse(getInboxTasksUri(), "mixed_inbox_tasks.json");
        mockWorkdayApiResponse(TIME_OFF_REQUEST_ID_1_API, "time_off_request_1.json");
        mockWorkdayApiResponse(TIME_OFF_REQUEST_ID_2_API, "time_off_request_2.json");

        final String actualResponseBody = requestCards(WORKDAY_AUTH_TOKEN, CARDS_REQUESTS_API,
                REQUEST_BODY_FILE_NAME, false)
            .expectStatus().isOk()
            .expectHeader().contentTypeCompatibleWith(APPLICATION_JSON)
            .returnResult(String.class)
            .getResponseBody()
            .collect(Collectors.joining())
            .block();

        assertThatJson(actualResponseBody).when(IGNORING_EXTRA_FIELDS, IGNORING_ARRAY_ORDER)
            .isEqualTo(fromFile("cards/cards.json"));
    }

    @Test public void canGetCardsWhenTimeOffDetailsResponseMissingTimeOffDescriptor() throws Exception {
        mockWorkdayApiResponse(getInboxTasksUri(), "inbox_task_timeOff.json");
        mockWorkdayApiResponse(TIME_OFF_REQUEST_ID_1_API, "time_off_response_missing_timeoff_descriptor.json");

        final String actualResponseBody = requestCards(WORKDAY_AUTH_TOKEN, CARDS_REQUESTS_API,
                REQUEST_BODY_FILE_NAME, false)
                .expectStatus().isOk()
                .expectHeader().contentTypeCompatibleWith(APPLICATION_JSON)
                .returnResult(String.class)
                .getResponseBody()
                .collect(Collectors.joining())
                .block();

        assertThatJson(actualResponseBody).when(IGNORING_EXTRA_FIELDS, IGNORING_ARRAY_ORDER)
                .isEqualTo(fromFile("cards/cards_without_type_attribute.json"));
    }

    @Test public void canGetCardsWithBusinessTitleChange() throws Exception {
        mockWorkdayApiResponse(getInboxTasksUri(), "mixed_inbox_tasks_with_businessTitleChange.json");
        mockWorkdayApiResponse(TIME_OFF_REQUEST_ID_1_API, "time_off_request_1.json");
        mockWorkdayApiResponse(TIME_OFF_REQUEST_ID_2_API, "time_off_request_2.json");
        mockBusinessTitleChangeDetails();

        final String actualResponseBody = requestCards(WORKDAY_AUTH_TOKEN, CARDS_REQUESTS_API, REQUEST_BODY_FILE_NAME, false)
                .expectStatus().isOk()
                .expectHeader().contentTypeCompatibleWith(APPLICATION_JSON)
                .returnResult(String.class)
                .getResponseBody()
                .collect(Collectors.joining())
                .block();

        assertThatJson(actualResponseBody).when(IGNORING_EXTRA_FIELDS, IGNORING_ARRAY_ORDER)
                .isEqualTo(fromFile("cards/cards_with_business_title_change.json"));
    }

    private void mockBusinessTitleChangeDetails() throws IOException {
        mockBackend.expect(requestTo(BUSINESS_TITLE_CHANGE_URL))
                .andExpect(header(ACCEPT, MediaType.APPLICATION_JSON_VALUE))
                .andExpect(header(AUTHORIZATION, WORKDAY_AUTH_TOKEN))
                .andExpect(method(GET))
                .andRespond(withSuccess(fromFile("business_title_change.json"), APPLICATION_JSON));
    }

    @Test public void canGetCardsWithBusinessProcess() throws Exception {
        mockWorkdayApiResponse(getInboxTasksUri(), "mixed_inbox_tasks_with_business_process.json");
        mockWorkdayApiResponse(TIME_OFF_REQUEST_ID_1_API, "time_off_request_1.json");
        mockWorkdayApiResponse(TIME_OFF_REQUEST_ID_2_API, "time_off_request_2.json");
        mockWorkdayApiResponse(BUSINESS_TITLE_CHANGE_URL, "business_title_change.json");
        mockWorkdayApiResponse(BUSINESS_PROCESS_URL, "Business_Process_Details.json");
        mockWorkdayApiResponse(BUSINESS_PROCESS_URL_1, "Business_Process_Details_1.json");
        mockWorkdayApiResponse(BUSINESS_PROCESS_URL_WITH_STEP_TYPE_ACTION, "Business_Process_Details_Step_Type_Action.json");
        mockWorkdayApiResponse(BUSINESS_PROCESS_URL_WITH_STEP_TYPE_REVIEW_DOCUMENTS, "Business_Process_Details_Step_Type_Review_Documents.json");

        final String actualResponseBody = requestCards(WORKDAY_AUTH_TOKEN, CARDS_REQUESTS_API, REQUEST_BODY_FILE_NAME, false)
                .expectStatus().isOk()
                .expectHeader().contentTypeCompatibleWith(APPLICATION_JSON)
                .returnResult(String.class)
                .getResponseBody()
                .collect(Collectors.joining())
                .block();

        assertThatJson(actualResponseBody).when(IGNORING_EXTRA_FIELDS, IGNORING_ARRAY_ORDER)
                .isEqualTo(fromFile("cards/cards_with_business_process.json"));
    }

    @Test public void canGetCardsWithBusinessProcessForPreHire() throws Exception {
        mockWorkdayApiResponse(getInboxTasksUri(), "inbox_task_pre_hire.json");
        mockWorkdayApiResponse(BUSINESS_PROCESS_URL_PRE_HIRE_1, "Business_Process_Details_Pre_Hire_1.json");
        mockWorkdayApiResponse(BUSINESS_PROCESS_URL_PRE_HIRE_2, "Business_Process_Details_Pre_Hire_2.json");

        final String actualResponseBody = requestCards(WORKDAY_AUTH_TOKEN, CARDS_REQUESTS_API, REQUEST_BODY_FILE_NAME, true)
                .expectStatus().isOk()
                .expectHeader().contentTypeCompatibleWith(APPLICATION_JSON)
                .returnResult(String.class)
                .getResponseBody()
                .collect(Collectors.joining())
                .block();

        assertThatJson(actualResponseBody).when(IGNORING_EXTRA_FIELDS, IGNORING_ARRAY_ORDER)
                .isEqualTo(fromFile("cards/cards_with_business_process_pre_hire.json"));
    }

    @Test public void testMissingRequestHeaders() throws IOException {
        String uri = "/cards/requests";
        webClient.post()
                .uri(uri)
                .header(AUTHORIZATION, "Bearer " + accessToken(uri))
                .contentType(APPLICATION_JSON)
                .accept(APPLICATION_JSON)
                .header(X_AUTH_HEADER, "Bearer abc")
                .bodyValue(fromFile(REQUEST_BODY_FILE_NAME))
                .exchange()
                .expectStatus().isBadRequest()
                .expectBody().jsonPath("$.message").isEqualTo("Missing request header 'X-Connector-Base-Url' for method parameter of type String");
    }

    @Test public void canGetCardsWhenInboxTaskHasDifferentDescriptor() throws Exception {
        mockWorkdayApiResponse(getInboxTasksUri(), "mixed_inbox_tasks_1.json");
        mockWorkdayApiResponse(TIME_OFF_REQUEST_ID_1_API, "time_off_request_diff_desc_1.json");
        mockWorkdayApiResponse(TIME_OFF_REQUEST_ID_2_API, "time_off_request_diff_desc_2.json");

        final String actualResponseBody = requestCards(WORKDAY_AUTH_TOKEN, CARDS_REQUESTS_API, REQUEST_BODY_FILE_NAME, false)
                .expectStatus().isOk()
                .expectHeader().contentTypeCompatibleWith(APPLICATION_JSON)
                .returnResult(String.class)
                .getResponseBody()
                .collect(Collectors.joining())
                .block();

        assertThatJson(actualResponseBody).when(IGNORING_EXTRA_FIELDS, IGNORING_ARRAY_ORDER)
                .isEqualTo(fromFile("cards/cards_diff_desc.json"));
    }
}

File: hub-ms-teams-connector/src/test/kotlin/com/vmware/connectors/msTeams/CardsControllerTest.kt
/*
* Copyright © 2020 VMware, Inc. All Rights Reserved.
* SPDX-License-Identifier: BSD-2-Clause
*/

package com.vmware.connectors.msTeams

import com.jayway.jsonpath.Configuration
import com.jayway.jsonpath.JsonPath
import com.jayway.jsonpath.Option
import com.jayway.jsonpath.spi.json.JacksonJsonNodeJsonProvider
import com.vmware.connectors.msTeams.config.Endpoints
import com.vmware.connectors.msTeams.config.ROUTING_PREFIX
import com.vmware.connectors.msTeams.utils.*
import com.vmware.connectors.test.ControllerTestsBase
import org.junit.Assert
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource
import org.springframework.beans.factory.annotation.Value
import org.springframework.core.io.Resource
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.http.MediaType
import org.springframework.test.web.client.ExpectedCount
import org.springframework.test.web.client.match.MockRestRequestMatchers
import org.springframework.test.web.client.response.MockRestResponseCreators
import org.springframework.test.web.reactive.server.expectBody
import org.springframework.test.web.reactive.server.returnResult
import org.springframework.util.LinkedMultiValueMap
import org.springframework.util.MimeTypeUtils
import org.springframework.web.reactive.function.BodyInserters
import org.springframework.web.reactive.function.client.WebClient
import uk.co.datumedge.hamcrest.json.SameJSONAs
import java.util.stream.Collectors

const val ROUTING_PREFIX_URL = "https://hero/connectors/ms_teams/"
const val messageId = "1570792618427"

class CardsControllerTest : ControllerTestsBase() {

    private fun getAuthorizationToken(uri: String) = "Bearer " + accessToken(uri)
    private val authorization = "Bearer abc"

    @ParameterizedTest
    @ValueSource(strings = [
        "/messages/$messageId/reply",
        "/messages/$messageId/dismiss",
        "/cards/requests"
    ])
    @Throws(Exception::class)
    fun testProtectedResource(uri: String) {
        testProtectedResource(HttpMethod.POST, uri)
    }

    @Test
    fun missingAuthorization() {
        webClient.post()
                .uri("/cards/requests")
                .exchange()
                .expectStatus().isUnauthorized
    }

    @Test
    @Throws(Exception::class)
    fun testGetImage(
            @Value("classpath:static/images/connector.png")
            icon: Resource
    ) {
        webClient.get()
                .uri("/images/connector.png")
                .exchange()
                .expectStatus().isOk
                .expectHeader().contentType(MimeTypeUtils.IMAGE_PNG_VALUE)
                .expectBody<ByteArray>().isEqualTo(icon.readAsByteArray())
    }

    @Test
    fun `Test Auth Fail For replyToMessage`(
            @Value("classpath:service/api/messages/message1.json")
            message: Resource
    ) {
        mockBackend.expect(MockRestRequestMatchers
                .requestTo(
                        Endpoints.getReplyToMessageUrl("28146088-6522-4459-9a52-295772efd51c",
                                "19:edc1a6c1be5c4736b7ead64dae9624e3@thread.skype",
                                "1570792618427", "")
                )
        )
                .andExpect(MockRestRequestMatchers.method(HttpMethod.POST))
                .andExpect(MockRestRequestMatchers.header(HttpHeaders.AUTHORIZATION, authorization))
                .andRespond(MockRestResponseCreators.withUnauthorizedRequest().body("Bad OAuth Token"))

        val formData = LinkedMultiValueMap<String, String>()
        formData.add("comments", "test subject")
        formData.add("actionType", "replyToMessage")
        formData.add("message", message.readAsString())

        val uri = "/messages/$messageId/reply"

        webClient.post()
                .uri(uri)
                .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                .header(HttpHeaders.AUTHORIZATION, getAuthorizationToken(uri))
                .header(X_AUTH_HEADER, authorization)
                .header(X_BASE_URL_HEADER, mockBackend.url(""))
                .body(BodyInserters.fromFormData(formData))
                .exchange()
                .expectStatus().isBadRequest
                .expectHeader().valueEquals("x-backend-status", "401")
    }

    @Test
    fun `Test Success For replyToMessage`(
            @Value("classpath:service/api/messages/message1.json")
            message: Resource
    ) {
        mockBackend.expect(MockRestRequestMatchers.requestTo(Endpoints.getReplyToMessageUrl("28146088-6522-4459-9a52-295772efd51c", "19:edc1a6c1be5c4736b7ead64dae9624e3@thread.skype", "1570792618427", "")))
                .andExpect(MockRestRequestMatchers.method(HttpMethod.POST))
                .andExpect(MockRestRequestMatchers.header(HttpHeaders.AUTHORIZATION, authorization))
                .andRespond(MockRestResponseCreators.withSuccess())

        val formData = LinkedMultiValueMap<String, String>()
        formData.add("comments", "test subject")
        formData.add("actionType", "replyToMessage")
        formData.add("message", message.readAsString())

        val uri = "/messages/$messageId/reply"

        webClient.post()
                .uri(uri)
                .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                .header(HttpHeaders.AUTHORIZATION, getAuthorizationToken(uri))
                .header(X_AUTH_HEADER, authorization)
                .header(X_BASE_URL_HEADER, mockBackend.url(""))
                .body(BodyInserters.fromFormData(formData))
                .exchange()
                .expectStatus().isOk

    }

    @Test
    fun `Test Success For replyToMessage With NullInput`(
            @Value("classpath:service/api/messages/message1.json")
            message: Resource
    ) {
        val formData = LinkedMultiValueMap<String, String>()
        formData.add("comments", null)
        formData.add("actionType", "replyToMessage")
        formData.add("message", message.readAsString())

        val uri = "/messages/$messageId/reply"

        webClient.post()
                .uri(uri)
                .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                .header(HttpHeaders.AUTHORIZATION, getAuthorizationToken(uri))
                .header(X_AUTH_HEADER, authorization)
                .header(X_BASE_URL_HEADER, mockBackend.url(""))
                .body(BodyInserters.fromFormData(formData))
                .exchange()
                .expectStatus().isOk
    }

    @Test
    fun `Test Success For Dismiss`(
            @Value("classpath:service/api/messages/message1.json")
            message: Resource
    ) {
        val formData = LinkedMultiValueMap<String, String>()
        formData.add("comments", "test subject")
        formData.add("actionType", "replyToMessage")
        formData.add("message", message.readAsString())

        val uri = "/messages/$messageId/dismiss"

        webClient.post()
                .uri(uri)
                .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                .header(HttpHeaders.AUTHORIZATION, getAuthorizationToken(uri))
                .header(X_AUTH_HEADER, authorization)
                .header(X_BASE_URL_HEADER, mockBackend.url(""))
                .body(BodyInserters.fromFormData(formData))
                .exchange()
                .expectStatus().isOk
    }

    @Test
    fun `Test Auth Fail For Cards Requests`(
            @Value("classpath:connector/responses/invalid_connector_token.json")
            response: Resource
    ) {
        mockBackend.expect(ExpectedCount.manyTimes(), MockRestRequestMatchers.requestTo(Endpoints.getTeamsUrl("")))
                .andExpect(MockRestRequestMatchers.method(HttpMethod.GET))
                .andExpect(MockRestRequestMatchers.header(HttpHeaders.AUTHORIZATION, authorization))
                .andRespond(MockRestResponseCreators.withUnauthorizedRequest())

        val jsonResponse = response.readAsString()

        val cardsUrl = "/cards/requests"

        webClient.post()
                .uri(cardsUrl)
                .header(HttpHeaders.AUTHORIZATION, getAuthorizationToken(cardsUrl))
                .header(X_AUTH_HEADER, authorization)
                .header(X_BASE_URL_HEADER, mockBackend.url(""))
                .header(ROUTING_PREFIX, ROUTING_PREFIX_URL)
                .contentType(MediaType.APPLICATION_JSON)
                .accept(MediaType.APPLICATION_JSON)
                .exchange()
                .expectStatus().isBadRequest
                .expectHeader().valueEquals("x-backend-status", "401")
                .expectHeader().contentTypeCompatibleWith(MediaType.APPLICATION_JSON)
                .expectBody().json(jsonResponse)

    }

    @Test
    fun `Test Success For Cards Request`(
            @Value("classpath:connector/responses/u1.json")
            response: Resource,
            @Value("classpath:service/api/teams/teams1.json")
            team: Resource,
            @Value("classpath:service/api/batchResponse/teamsBatchResponse1.json")
            teamsBatchResponse: Resource,
            @Value("classpath:service/api/batchResponse/channelBatchResponse1.json")
            channelBatchResponse: Resource,
            @Value("classpath:service/api/batchResponse/replyBatchResponse1.json")
            replyBatchResponse: Resource
    ) {
        mockBackend.expect(MockRestRequestMatchers.requestTo(Endpoints.getTeamsUrl("")))
                .andExpect(MockRestRequestMatchers.method(HttpMethod.GET))
                .andExpect(MockRestRequestMatchers.header(HttpHeaders.AUTHORIZATION, authorization))
                .andRespond(MockRestResponseCreators.withSuccess(team.readAsString(), MediaType.APPLICATION_JSON))

        mockBackend.expect(MockRestRequestMatchers.requestTo(Endpoints.getBatchUrl("")))
                .andExpect(MockRestRequestMatchers.method(HttpMethod.POST))
                .andExpect(MockRestRequestMatchers.header(HttpHeaders.AUTHORIZATION, authorization))
                .andExpect(MockRestRequestMatchers.header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE))
                .andRespond(MockRestResponseCreators.withSuccess(teamsBatchResponse.readAsString(), MediaType.APPLICATION_JSON))

        (0..4).map {
            mockBackend.expect(MockRestRequestMatchers.requestTo(Endpoints.getBatchUrl("")))
                    .andExpect(MockRestRequestMatchers.method(HttpMethod.POST))
                    .andExpect(MockRestRequestMatchers.header(HttpHeaders.AUTHORIZATION, authorization))
                    .andExpect(MockRestRequestMatchers.header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE))
                    .andRespond(MockRestResponseCreators.withSuccess(channelBatchResponse.readAsString(), MediaType.APPLICATION_JSON))
        }
        (0..1).map {
            mockBackend.expect(MockRestRequestMatchers.requestTo(Endpoints.getBatchUrl("")))
                    .andExpect(MockRestRequestMatchers.method(HttpMethod.POST))
                    .andExpect(MockRestRequestMatchers.header(HttpHeaders.AUTHORIZATION, authorization))
                    .andExpect(MockRestRequestMatchers.header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE))
                    .andRespond(MockRestResponseCreators.withSuccess(replyBatchResponse.readAsString(), MediaType.APPLICATION_JSON))
        }
        val jsonResponse = response.readAsString()

        val uri = "/cards/requests"

        val data = webClient.post()
                .uri(uri)
                .header(HttpHeaders.AUTHORIZATION, getAuthorizationToken(uri))
                .header(X_AUTH_HEADER, authorization)
                .header(X_BASE_URL_HEADER, mockBackend.url(""))
                .header(ROUTING_PREFIX, ROUTING_PREFIX_URL)
                .contentType(MediaType.APPLICATION_JSON)
                .accept(MediaType.APPLICATION_JSON)
                .exchange()
                .expectStatus().isOk
                .expectHeader().contentTypeCompatibleWith(MediaType.APPLICATION_JSON)
                .returnResult<String>()
                .responseBody
                .collect(Collectors.joining())
                .map(this::normalizeCards)
                .block()
                ?.replace(Regex("http://localhost:\\d+/"), "/")
        Assert.assertThat<String>(data, SameJSONAs.sameJSONAs(jsonResponse))
    }

    @Test
    fun `Test Success For Empty Cards`(
            @Value("classpath:connector/responses/u5.json")
            response: Resource,
            @Value("classpath:service/api/teams/teams1.json")
            team: Resource,
            @Value("classpath:service/api/batchResponse/teamsBatchResponse1.json")
            teamsBatchResponse: Resource,
            @Value("classpath:service/api/batchResponse/channelBatchResponse1.json")
            channelBatchResponse: Resource,
            @Value("classpath:service/api/batchResponse/channelBatchResponse2.json")
            channelBatchResponse1: Resource,
            @Value("classpath:service/api/batchResponse/replyBatchResponse2.json")
            replyBatchResponse: Resource
    ) {
        mockBackend.expect(MockRestRequestMatchers.requestTo(Endpoints.getTeamsUrl("")))
                .andExpect(MockRestRequestMatchers.method(HttpMethod.GET))
                .andExpect(MockRestRequestMatchers.header(HttpHeaders.AUTHORIZATION, authorization))
                .andRespond(MockRestResponseCreators.withSuccess(team.readAsString(), MediaType.APPLICATION_JSON))

        mockBackend.expect(MockRestRequestMatchers.requestTo(Endpoints.getBatchUrl("")))
                .andExpect(MockRestRequestMatchers.method(HttpMethod.POST))
                .andExpect(MockRestRequestMatchers.header(HttpHeaders.AUTHORIZATION, authorization))
                .andExpect(MockRestRequestMatchers.header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE))
                .andRespond(MockRestResponseCreators.withSuccess(teamsBatchResponse.readAsString(), MediaType.APPLICATION_JSON))

        mockBackend.expect(MockRestRequestMatchers.requestTo(Endpoints.getBatchUrl("")))
                .andExpect(MockRestRequestMatchers.method(HttpMethod.POST))
                .andExpect(MockRestRequestMatchers.header(HttpHeaders.AUTHORIZATION, authorization))
                .andExpect(MockRestRequestMatchers.header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE))
                .andRespond(MockRestResponseCreators.withSuccess(channelBatchResponse1.readAsString(), MediaType.APPLICATION_JSON))

        (0..4).map {
            mockBackend.expect(MockRestRequestMatchers.requestTo(Endpoints.getBatchUrl("")))
                    .andExpect(MockRestRequestMatchers.method(HttpMethod.POST))
                    .andExpect(MockRestRequestMatchers.header(HttpHeaders.AUTHORIZATION, authorization))
                    .andExpect(MockRestRequestMatchers.header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE))
                    .andRespond(MockRestResponseCreators.withSuccess(channelBatchResponse.readAsString(), MediaType.APPLICATION_JSON))
        }
        mockBackend.expect(ExpectedCount.manyTimes(), MockRestRequestMatchers.requestTo(Endpoints.getBatchUrl("")))
                .andExpect(MockRestRequestMatchers.method(HttpMethod.POST))
                .andExpect(MockRestRequestMatchers.header(HttpHeaders.AUTHORIZATION, authorization))
                .andExpect(MockRestRequestMatchers.header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE))
                .andRespond(MockRestResponseCreators.withSuccess(replyBatchResponse.readAsString(), MediaType.APPLICATION_JSON))

        val jsonResponse = response.readAsString()

        val uri = "/cards/requests"

        val data = webClient.post()
                .uri(uri)
                .header(HttpHeaders.AUTHORIZATION, getAuthorizationToken(uri))
                .header(X_AUTH_HEADER, authorization)
                .header(X_BASE_URL_HEADER, mockBackend.url(""))
                .header(ROUTING_PREFIX, ROUTING_PREFIX_URL)
                .contentType(MediaType.APPLICATION_JSON)
                .accept(MediaType.APPLICATION_JSON)
                .exchange()
                .expectStatus().isOk
                .expectHeader().contentTypeCompatibleWith(MediaType.APPLICATION_JSON)
                .returnResult<String>()
                .responseBody
                .collect(Collectors.joining())
                .map(this::normalizeCards)
                .block()
                ?.replace(Regex("http://localhost:\\d+/"), "/")

        Assert.assertThat<String>(data, SameJSONAs.sameJSONAs(jsonResponse))
    }

    @Test
    fun `Test Auth Fail While Batch Request`(
            @Value("classpath:connector/responses/invalid_connector_token.json")
            response: Resource,
            @Value("classpath:service/api/teams/teams1.json")
            team: Resource
    ) {
        mockBackend.expect(MockRestRequestMatchers.requestTo(Endpoints.getTeamsUrl("")))
                .andExpect(MockRestRequestMatchers.method(HttpMethod.GET))
                .andExpect(MockRestRequestMatchers.header(HttpHeaders.AUTHORIZATION, authorization))
                .andRespond(MockRestResponseCreators.withSuccess(team.readAsString(), MediaType.APPLICATION_JSON))

        mockBackend.expect(MockRestRequestMatchers.requestTo(Endpoints.getBatchUrl("")))
                .andExpect(MockRestRequestMatchers.method(HttpMethod.POST))
                .andExpect(MockRestRequestMatchers.header(HttpHeaders.AUTHORIZATION, authorization))
                .andExpect(MockRestRequestMatchers.header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE))
                .andRespond(MockRestResponseCreators.withUnauthorizedRequest())

        val jsonResponse = response.readAsString()

        val uri = "/cards/requests"

        webClient.post()
                .uri(uri)
                .header(HttpHeaders.AUTHORIZATION, getAuthorizationToken(uri))
                .header(X_AUTH_HEADER, authorization)
                .header(X_BASE_URL_HEADER, mockBackend.url(""))
                .header(ROUTING_PREFIX, ROUTING_PREFIX_URL)
                .contentType(MediaType.APPLICATION_JSON)
                .accept(MediaType.APPLICATION_JSON)
                .exchange()
                .expectStatus()
                .isBadRequest
                .expectHeader().valueEquals("x-backend-status", "401")
                .expectHeader().contentTypeCompatibleWith(MediaType.APPLICATION_JSON)
                .expectBody().json(jsonResponse)
    }

    @Test
    fun miscellaneous() {
        val map = mapOf("string" to null)
        val resp = map.getListOrDefault("s", listOf(1, 2))
        Assert.assertEquals(listOf(1, 2), resp)
        val mapResp = map.getMapOrDefault("s", mapOf("sample" to 1))
        Assert.assertEquals(mapOf("sample" to 1), mapResp)
        val emptyMapResp = map.getMapOrDefault<String, Any>("s")
        Assert.assertEquals(emptyMap<String, Any>(), emptyMapResp)
        Assert.assertEquals(null, map.getMapOrNull<String, Any>("j"))
        try {
            map.getListOrException<String>("f")
        } catch (e: java.lang.Exception) {
            Assert.assertEquals(null, e.localizedMessage)
        }
        try {
            map.getMapOrException<String, Any>("f")
        } catch (e: java.lang.Exception) {
            Assert.assertEquals(null, e.localizedMessage)
        }
        val userNameFromToken = VmwareUtils.getUserEmailFromToken("token s.bnVsbA==")
        Assert.assertEquals(null, userNameFromToken)
    }

    fun normalizeCards(body: String?): String? {
        val configuration = Configuration.builder()
                .options(Option.SUPPRESS_EXCEPTIONS)
                .jsonProvider(JacksonJsonNodeJsonProvider())
                .build()
        val context = JsonPath.using(configuration).parse(body)
        context.set("$.objects[*].id", "00000000-0000-0000-0000-000000000000")
        context.set("$.objects[*].creation_date", "1970-01-01T00:00:00Z")
        context.set("$.objects[*].expiration_date", "1970-01-01T00:00:00Z")
        context.set("$.objects[*].actions[*].id", "00000000-0000-0000-0000-000000000000")
        return context.jsonString()
    }
}
